### PR TITLE
New frontend handling of supervision disaggregations

### LINKED ISSUE #321: Frontend changes to go along with "Change how supervision disaggregation works" #17802
This task is to make the frontend changes that go along with [this backend PR](https://github.com/Recidiviz/recidiviz-data/pull/17802).

Do the following:
1) Read that backend PR description and the actual code, to make sure you understand what changes are being made.
2) Check out that PR and run it locally. Go to a Supervision system. Make sure you've selected some supervision subsystems in Agency Settings, and then and try toggling on/off the disaggregated-by-supervision-subsystems flag for some metrics. Make sure you can see how (after a manual refresh) the enabled/disabled statuses are changing.
3) Think about what the minimal frontend changes we need to support this change are (i.e. remove the toast and redirect, update the enabled/disabled statuses in place without a manual refresh).
4) Think about what the *best* frontend experience would be. See Humphrey's comments in Slack [here](https://recidiviz.slack.com/archives/C02VDN1LSSW/p1674025862137109?thread_ts=1673993096.344509&cid=C02VDN1LSSW).

### PR DESCRIPTION
**Pull request by terryttsai**
## Description of the change

https://www.loom.com/share/9813a11e0a424d1cafcd11db85fb6cc2

Went through this with Humphrey: now that we see all supervision disaggregated metrics, we simply enable/disable them when they are toggled between combined and disaggregated.

This is enough to avoid the refresh experience. There is an even bigger design exercise in the future to make this more intuitive.

## Type of change

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #321

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/components/MetricConfiguration/Configuration.tsx [hunk 1]
```diff
@@ -144,34 +147,23 @@ export const Configuration: React.FC<MetricConfigurationProps> = observer(
...
 
-        setTimeout(
-          () => showToast({ message: toastMessage, timeout: 5000 }),
-          1000
-        );
-        setTimeout(() => {
-          navigate("../metric-config");
-          window.location.reload();
-        }, 5000);
+        // After saving disaggregation selection, re-fetch metric settings
```
Comment (lilidworkin) [inline publisher/src/components/MetricConfiguration/Configuration.tsx:163]: Oh interesting way to solve this! I'm fine with this, but I'm wondering, what backend returned the new metric config values in the saveMetricSettings response? That way we'd avoid another round trip? I think we could make that change. (Would we need to return all of them, or could we just return the ones that changed?)

File: publisher/src/components/MetricConfiguration/Configuration.tsx [hunk 2]
```diff
@@ -239,177 +231,191 @@ export const Configuration: React.FC<MetricConfigurationProps> = observer(
...
-                <p>
-                  Disaggregations include the populations you selected in{" "}
-                  <BlueLinkSpan onClick={() => navigate("../agency-settings")}>
-                    Agency Settings
-                  </BlueLinkSpan>{" "}
-                  ({printCommaSeparatedList(capitalizedSupervisionSubsystems)}
-                  ).
-                </p>
-                <p>
-                  NOTE: Changing this option will refresh the page to reflect
```
Comment (lilidworkin) [inline publisher/src/components/MetricConfiguration/Configuration.tsx:377]: Do we still want to include a note explaining how this toggle will enable/disable this metric and enable/disable the other metrics in the other supervision subpopulations? Might be nice. cc @hobuobi

**Review by lilidworkin — APPROVED** [review]
@terryttsai this looks awesome! This is *exactly* what I had in mind. Just left two questions inline. I think regardless we should merge this in and can address the other items in a followup PR!

File: publisher/src/components/MetricConfiguration/Configuration.tsx (hunk 2 above)
Comment (terryttsai) [reply publisher/src/components/MetricConfiguration/Configuration.tsx:377]: Down if that helps! I'm not sure if "Supervision subpopulation" maps to "Revocation (Parole)", "Revocation (Probation)" etc on its own or if users would already understand that

File: publisher/src/components/MetricConfiguration/Configuration.tsx (hunk 1 above)
Comment (terryttsai) [reply publisher/src/components/MetricConfiguration/Configuration.tsx:163]: Since this whole experience is being redesigned anyways, I don't know if it's worth putting much more effort into making this cleaner. 

If we did want to make this cleaner, I was thinking if there's someway for the front-end to know how to map between a metric's combined and disaggregated versions, then the front-end can just modify the mobx store accordingly, without needing any additional information from the backend

Comment (mxosman) [reply publisher/src/components/MetricConfiguration/Configuration.tsx:163]: Ah - this is a super clever solution! Love both ideas here as well as far as reducing the # of round trip API calls.

File: publisher/src/components/MetricConfiguration/Configuration.tsx [hunk 3]
```diff
@@ -239,177 +231,191 @@ export const Configuration: React.FC<MetricConfigurationProps> = observer(
...
-          )}
+          <TogglableSection
+            enabled={
+              !(
+                (disaggregatedBySupervisionSubsystems === true &&
+                  systemSearchParam === SupervisionSystem) ||
+                (disaggregatedBySupervisionSubsystems === false &&
+                  systemSearchParam &&
+                  SupervisionSubsystems.includes(systemSearchParam))
+              )
```
Comment (mxosman) [inline publisher/src/components/MetricConfiguration/Configuration.tsx:242]: super tiny nit & only if you feel moved: maybe worth putting this in a descriptive variable?

File: common/types.ts [hunk 4]
```diff
@@ -40,8 +40,9 @@ export type AgencyTeam = {
   email: string;
 };
 
-export const SupervisionSystems: AgencySystems[] = [
-  "SUPERVISION",
+export const SupervisionSystem: AgencySystems = "SUPERVISION";
+
+export const SupervisionSubsystems: AgencySystems[] = [
```
Comment (mxosman) [inline common/types.ts:45]: Love the way you broke these two apart!

**Review by mxosman — APPROVED** [review]

**Comment by mxosman** [comment]
Awesome solution here Terry & Lili!

File: publisher/src/components/MetricConfiguration/Configuration.tsx (hunk 3 above)
Comment (terryttsai) [reply publisher/src/components/MetricConfiguration/Configuration.tsx:242]: Good idea, it's a bit confusing!

File: publisher/src/components/MetricConfiguration/Configuration.tsx (hunk 2 above)
Comment (lilidworkin) [reply publisher/src/components/MetricConfiguration/Configuration.tsx:377]: Hm yea, maybe let's discuss this in playtesting tomorrow and see if folks ahve thoughts!

File: publisher/src/components/MetricConfiguration/Configuration.tsx (hunk 1 above)
Comment (lilidworkin) [reply publisher/src/components/MetricConfiguration/Configuration.tsx:163]: > If we did want to make this cleaner, I was thinking if there's someway for the front-end to know how to map between a metric's combined and disaggregated versions, then the front-end can just modify the mobx store accordingly, without needing any additional information from the backend

@terryttsai yes! There should be a way to do this, indeed. Something like this:
Aggregated:
SUPERVISION_STAFF enabled
PAROLE_STAFF disabled
PROBATION_STAFF disabled

Disaggregated: reverse

Basically, the metric key is SYSTEM_<METRIC_NAME>. The SYSTEMs of interest are SUPERVISION and then whichever other ones the agency has specified (which we should be already surfacing in the UI here, @mxosman would know about that). We should be able to determine the other metric keys by just doing a string replacement. LMK if that makes sense!

**Comment by lilidworkin** [comment]
@terryttsai is it okay if I merge this so I can get it into a deploy for playtesting? We can address any followups in a separate PR!

**Comment by terryttsai** [comment]
I'll merge this in for playtesting now!
